File: 1461-check-if-a-string-contains-all-binary-codes-of-size-k/1461-check-if-a-string-contains-all-binary-codes-of-size-k.cpp
#include <vector>

class Solution {
public:
    bool hasAllCodes(string s, int k) {
        // rolling hash
        vector<bool> found(1 << k);
        int num = 0, mask = (1 << k) - 1, cnt = 0;
        for (int i = 0; i < s.size(); ++i) {
            num = mask & ((num << 1) + s[i] - '0');
            if (i >= k - 1) {
                if (found[num] == false)
                    ++cnt;
                found[num] = true;
            }
        }
        return cnt == found.size();
    }
};

File: 1461-check-if-a-string-contains-all-binary-codes-of-size-k/1461-check-if-a-string-contains-all-binary-codes-of-size-k.java
class Solution {
    public boolean hasAllCodes(String s, int k) {
        boolean[] found = new boolean[1 << k];
        int num = 0, mask = (1 << k) - 1, cnt = 0;
        char[] ca = s.toCharArray();
        for (int i = 0; i < ca.length; ++i) {
            num = mask & ((num << 1) + ca[i] - '0');
            if (i >= k - 1) {
                if (found[num] == false)
                    ++cnt;
                found[num] = true;
            }
        }
        return cnt == found.length;
    }
}